File: packages/go/cairn-p2p/node.go
package cairn

import (
	"context"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"sync"
)

// NetworkInfo holds read-only diagnostic information about the node's network state.
type NetworkInfo struct {
	PeerID    PeerID
	NatType   string
	Addresses []string
}

// PairingHandler is an interface for pluggable pairing logic.
// It allows the Node to delegate pairing operations without importing crypto packages.
type PairingHandler interface {
	GenerateQR(peerID PeerID) ([]byte, error)
	ScanQR(data []byte) (PeerID, error)
	GeneratePin() (string, error)
	EnterPin(pin string) (PeerID, error)
	GenerateLink(peerID PeerID) (string, error)
	FromLink(uri string) (PeerID, error)
	Pair(peerID PeerID, method PairingMethod) (PeerID, error)
}

// ConnectHandler is an interface for pluggable connection logic.
// It performs the Noise XX handshake and Double Ratchet initialization.
type ConnectHandler interface {
	Connect(peerID PeerID) (*Encryptor, error)
}

// Encryptor wraps Double Ratchet encrypt/decrypt behind an interface
// to avoid import cycles with the crypto package.
type Encryptor struct {
	Encrypt    func(plaintext []byte) (header []byte, ciphertext []byte, err error)
	Decrypt    func(header []byte, ciphertext []byte) ([]byte, error)
	CloseFunc  func()
}

// CustomMessageHandler is a callback for node-level custom message handling.
type CustomMessageHandler func(peerID PeerID, data []byte)

// Node is the primary entry point for the cairn P2P connectivity library.
// It manages identity, pairing, sessions, discovery, and transport.
type Node struct {
	mu sync.RWMutex

	config      *Config
	peerID      PeerID
	publicKey   ed25519.PublicKey
	identitySeed [32]byte // Ed25519 seed for identity reconstruction
	events      chan Event
	sessions    map[PeerID]*Session

	pairedPeers    map[PeerID]bool
	pairingHandler PairingHandler
	connectHandler ConnectHandler

	customRegistry map[uint16]CustomMessageHandler
	customMu       sync.RWMutex

	// Transport state (populated after StartTransport)
	listenAddresses []string
	transportReady  bool

	ctx    context.Context
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

// newNode creates a Node with the given config. Internal constructor.
func newNode(config *Config) (*Node, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate identity: %w", err)
	}

	// Save the 32-byte seed for identity reconstruction by handler packages.
	var seed [32]byte
	copy(seed[:], priv.Seed())

	// Derive PeerID from public key (multihash: 0x12, 0x20, SHA-256)
	hash := sha256.Sum256(pub)
	var pid PeerID
	pid[0] = 0x12
	pid[1] = 0x20
	copy(pid[2:], hash[:])

	ctx, cancel := context.WithCancel(context.Background())

	node := &Node{
		config:         config,
		peerID:         pid,
		publicKey:      pub,
		identitySeed:   seed,
		events:         make(chan Event, config.EventBufferSize),
		sessions:       make(map[PeerID]*Session),
		pairedPeers:    make(map[PeerID]bool),
		customRegistry: make(map[uint16]CustomMessageHandler),
		ctx:            ctx,
		cancel:         cancel,
	}

	// Apply default wiring if registered (by importing the api package).
	if defaultNodeWiring != nil {
		if err := defaultNodeWiring(node); err != nil {
			cancel()
			return nil, fmt.Errorf("default wiring: %w", err)
		}
	}

	return node, nil
}

// Events returns the read-only event channel for this node.
// Events include StateChangedEvent, MessageReceivedEvent, ChannelOpenedEvent,
// and PeerUnpairedEvent.
func (n *Node) Events() <-chan Event {
	return n.events
}

// PeerID returns this node's peer ID.
func (n *Node) PeerID() PeerID {
	return n.peerID
}

// IdentitySeed returns the 32-byte Ed25519 seed for this node's identity.
// Used by handler packages to reconstruct the full identity keypair.
func (n *Node) IdentitySeed() [32]byte {
	return n.identitySeed
}

// SetPairingHandler sets the pairing handler for this node.
func (n *Node) SetPairingHandler(h PairingHandler) {
	n.mu.Lock()
	defer n.mu.Unlock()
	n.pairingHandler = h
}

// SetConnectHandler sets the connection handler for this node.
func (n *Node) SetConnectHandler(h ConnectHandler) {
	n.mu.Lock()
	defer n.mu.Unlock()
	n.connectHandler = h
}

// RegisterCustomMessage registers a node-wide handler for a custom message type (0xF000-0xFFFF).
//
// Node-level handlers are invoked when a custom message arrives on any session
// that does not have a per-session handler for the type code.
func (n *Node) RegisterCustomMessage(typeCode uint16, handler CustomMessageHandler) error {
	if typeCode < 0xF000 || typeCode > 0xFFFF {
		return fmt.Errorf("custom message type 0x%04X outside application range 0xF000-0xFFFF", typeCode)
	}
	n.customMu.Lock()
	defer n.customMu.Unlock()
	n.customRegistry[typeCode] = handler
	return nil
}

// Connect initiates or resumes a session with a paired peer.
func (n *Node) Connect(ctx context.Context, peerID PeerID) (*Session, error) {
	n.mu.Lock()
	defer n.mu.Unlock()

	if s, ok := n.sessions[peerID]; ok {
		if s.State() != StateFailed {
			return s, nil
		}
	}

	session := NewSession(peerID, &SessionConfig{
		SessionExpiry:          n.config.ReconnectionPolicy.SessionExpiry,
		HeartbeatInterval:      n.config.ReconnectionPolicy.HeartbeatInterval,
		HeartbeatTimeout:       n.config.ReconnectionPolicy.HeartbeatTimeout,
		ReconnectInitialDelay:  n.config.ReconnectionPolicy.BackoffInitial,
		ReconnectMaxDelay:      n.config.ReconnectionPolicy.BackoffMax,
		ReconnectBackoffFactor: n.config.ReconnectionPolicy.BackoffFactor,
	}, n.events)

	// If a connect handler is set, perform handshake and set up encryption
	if n.connectHandler != nil {
		enc, err := n.connectHandler.Connect(peerID)
		if err != nil {
			return nil, fmt.Errorf("connect handshake: %w", err)
		}
		if enc != nil {
			session.SetEncryptor(enc)
		}
	}

	n.sessions[peerID] = session
	return session, nil
}

// Unpair removes a paired peer and notifies the remote peer.
func (n *Node) Unpair(ctx context.Context, peerID PeerID) error {
	n.mu.Lock()
	defer n.mu.Unlock()

	if s, ok := n.sessions[peerID]; ok {
		s.Close()
		delete(n.sessions, peerID)
	}
	delete(n.pairedPeers, peerID)

	n.events <- PeerUnpairedEvent{PeerID: peerID}
	return nil
}

// StartTransport initializes the transport layer (TCP listener on an
// ephemeral port). After this call, Connect() can dial peers over the
// real network. Safe to skip in unit tests.
func (n *Node) StartTransport() error {
	n.mu.Lock()
	defer n.mu.Unlock()

	// For Go, we use the TransportChain's TCP provider.
	// The TransportChain itself is defined in transport/chain.go
	// For now, mark transport as ready and record a placeholder address.
	// Full libp2p integration (like Rust) will follow in a future PR.
	n.transportReady = true
	n.listenAddresses = []string{fmt.Sprintf("/ip4/0.0.0.0/tcp/0/p2p/%s", n.peerID)}
	return nil
}

// ListenAddresses returns the node's listen addresses (available after StartTransport).
func (n *Node) ListenAddresses() []string {
	n.mu.RLock()
	defer n.mu.RUnlock()
	addrs := make([]string, len(n.listenAddresses))
	copy(addrs, n.listenAddresses)
	return addrs
}

// TransportReady returns whether the transport layer has been started.
func (n *Node) TransportReady() bool {
	n.mu.RLock()
	defer n.mu.RUnlock()
	return n.transportReady
}

// NetworkInfo returns read-only diagnostic information about the node's network.
func (n *Node) NetworkInfo() NetworkInfo {
	n.mu.RLock()
	defer n.mu.RUnlock()
	return NetworkInfo{
		PeerID:    n.peerID,
		NatType:   "unknown",
		Addresses: n.listenAddresses,
	}
}

// Config returns the node's configuration.
func (n *Node) Config() *Config {
	return n.config
}

// Close gracefully shuts down the node, cancelling all operations and
// waiting for goroutines to complete.
func (n *Node) Close() error {
	n.cancel()

	n.mu.Lock()
	for _, s := range n.sessions {
		s.Close()
	}
	n.sessions = make(map[PeerID]*Session)
	n.mu.Unlock()

	n.wg.Wait()
	close(n.events)
	return nil
}

// registerPairedPeer adds a peer to the paired set.
func (n *Node) registerPairedPeer(peerID PeerID) {
	n.pairedPeers[peerID] = true
}

// emitPairingComplete sends a PairingCompleteEvent on the events channel.
// Must be called without holding n.mu to avoid blocking.
func (n *Node) emitPairingComplete(peerID PeerID, method PairingMethod) {
	select {
	case n.events <- PairingCompleteEvent{PeerID: peerID, Method: method}:
	default:
		// Drop if channel full (non-blocking to avoid deadlock).
	}
}

// --- Pairing methods ---
// These delegate to the PairingHandler if set, otherwise return stubs.

// PairGenerateQR generates a QR code pairing payload.
func (n *Node) PairGenerateQR(ctx context.Context) ([]byte, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		return h.GenerateQR(n.peerID)
	}
	return nil, fmt.Errorf("not yet integrated with transport layer")
}

// PairScanQR processes a scanned QR code and initiates pairing.
func (n *Node) PairScanQR(ctx context.Context, data []byte) (PeerID, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		pid, err := h.ScanQR(data)
		if err != nil {
			return PeerID{}, err
		}
		n.mu.Lock()
		n.registerPairedPeer(pid)
		n.mu.Unlock()
		n.emitPairingComplete(pid, PairingQR)
		return pid, nil
	}
	return PeerID{}, fmt.Errorf("not yet integrated with transport layer")
}

// PairGeneratePin generates a PIN code for pairing.
func (n *Node) PairGeneratePin(ctx context.Context) (string, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		return h.GeneratePin()
	}
	return "", fmt.Errorf("not yet integrated with transport layer")
}

// PairEnterPin enters a PIN code to pair with a remote peer.
func (n *Node) PairEnterPin(ctx context.Context, pin string) (PeerID, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		pid, err := h.EnterPin(pin)
		if err != nil {
			return PeerID{}, err
		}
		n.mu.Lock()
		n.registerPairedPeer(pid)
		n.mu.Unlock()
		n.emitPairingComplete(pid, PairingPin)
		return pid, nil
	}
	return PeerID{}, fmt.Errorf("not yet integrated with transport layer")
}

// PairGenerateLink generates a pairing link URI.
func (n *Node) PairGenerateLink(ctx context.Context) (string, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		return h.GenerateLink(n.peerID)
	}
	return "", fmt.Errorf("not yet integrated with transport layer")
}

// PairFromLink initiates pairing from a pairing link URI.
func (n *Node) PairFromLink(ctx context.Context, uri string) (PeerID, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		pid, err := h.FromLink(uri)
		if err != nil {
			return PeerID{}, err
		}
		n.mu.Lock()
		n.registerPairedPeer(pid)
		n.mu.Unlock()
		n.emitPairingComplete(pid, PairingLink)
		return pid, nil
	}
	return PeerID{}, fmt.Errorf("not yet integrated with transport layer")
}

// Pair initiates pairing with a specific peer using the given method.
func (n *Node) Pair(ctx context.Context, peerID PeerID, method PairingMethod) (PeerID, error) {
	n.mu.RLock()
	h := n.pairingHandler
	n.mu.RUnlock()
	if h != nil {
		pid, err := h.Pair(peerID, method)
		if err != nil {
			return PeerID{}, err
		}
		n.mu.Lock()
		n.registerPairedPeer(pid)
		n.mu.Unlock()
		n.emitPairingComplete(pid, method)
		return pid, nil
	}
	return PeerID{}, fmt.Errorf("not yet integrated with transport layer")
}

// --- Envelope helpers ---

// ChannelInitPayload is the CBOR payload for a channel init message.
type ChannelInitPayload struct {
	Name string `cbor:"0,keyasint"`
}

// MsgChannelInit is the message type for channel initialization.
const MsgChannelInit uint16 = 0x0303

// createDataEnvelope creates a CBOR-encoded MessageEnvelope for a data message.
// This is defined in the root package to avoid import cycles.
func createDataEnvelope(sessionID [16]byte, seqNum uint64, payload []byte) ([]byte, error) {
	// Manual CBOR map encoding to avoid importing protocol package
	// Format: CBOR map with integer keys {0: version, 1: type, 2: msgId, 3: sessionId, 4: payload, 5: authTag}
	return encodeEnvelope(1, 0x0300, sessionID, payload, seqNum)
}

// createChannelInitEnvelope creates a CBOR-encoded envelope for a channel init message.
func createChannelInitEnvelope(sessionID [16]byte, channelName string) ([]byte, error) {
	// Encode the ChannelInitPayload as CBOR
	initPayload := cborEncodeChannelInit(channelName)
	return encodeEnvelope(1, MsgChannelInit, sessionID, initPayload, 0)
}

// encodeEnvelope manually constructs a minimal CBOR-encoded envelope.
// This avoids importing the protocol package which would create an import cycle.
func encodeEnvelope(version uint8, msgType uint16, sessionID [16]byte, payload []byte, seqNum uint64) ([]byte, error) {
	// We use a simple binary format: version(1) + type(2) + sessionID(16) + seqNum(8) + payloadLen(4) + payload
	buf := make([]byte, 0, 1+2+16+8+4+len(payload))
	buf = append(buf, version)
	buf = binary.BigEndian.AppendUint16(buf, msgType)
	buf = append(buf, sessionID[:]...)
	buf = binary.BigEndian.AppendUint64(buf, seqNum)
	buf = binary.BigEndian.AppendUint32(buf, uint32(len(payload)))
	buf = append(buf, payload...)
	return buf, nil
}

// decodeEnvelopeSimple decodes the simple binary envelope format.
func decodeEnvelopeSimple(data []byte) (version uint8, msgType uint16, sessionID [16]byte, seqNum uint64, payload []byte, err error) {
	if len(data) < 1+2+16+8+4 {
		return 0, 0, [16]byte{}, 0, nil, fmt.Errorf("envelope too short: %d bytes", len(data))
	}
	version = data[0]
	msgType = binary.BigEndian.Uint16(data[1:3])
	copy(sessionID[:], data[3:19])
	seqNum = binary.BigEndian.Uint64(data[19:27])
	payloadLen := binary.BigEndian.Uint32(data[27:31])
	if uint32(len(data)-31) < payloadLen {
		return 0, 0, [16]byte{}, 0, nil, fmt.Errorf("envelope payload truncated")
	}
	payload = data[31 : 31+payloadLen]
	return version, msgType, sessionID, seqNum, payload, nil
}

// cborEncodeChannelInit encodes a channel name as a minimal CBOR map {0: name}.
func cborEncodeChannelInit(name string) []byte {
	// CBOR: map(1) { 0: text(name) }
	// A1 00 [text string]
	nameBytes := []byte(name)
	buf := make([]byte, 0, 3+len(nameBytes))
	buf = append(buf, 0xA1) // map of 1 pair
	buf = append(buf, 0x00) // key: 0
	if len(nameBytes) < 24 {
		buf = append(buf, 0x60+byte(len(nameBytes))) // text string, length < 24
	} else {
		buf = append(buf, 0x78, byte(len(nameBytes))) // text string, 1-byte length
	}
	buf = append(buf, nameBytes...)
	return buf
}

// cborDecodeChannelInit decodes a channel name from the minimal CBOR map.
func cborDecodeChannelInit(data []byte) (string, error) {
	if len(data) < 3 {
		return "", fmt.Errorf("channel init payload too short")
	}
	if data[0] != 0xA1 || data[1] != 0x00 {
		return "", fmt.Errorf("unexpected channel init CBOR structure")
	}
	major := data[2] & 0xE0
	if major != 0x60 { // major type 3 = text string
		return "", fmt.Errorf("expected text string for channel name")
	}
	nameLen := int(data[2] & 0x1F)
	if nameLen == 24 {
		if len(data) < 4 {
			return "", fmt.Errorf("channel init payload too short for 1-byte length")
		}
		nameLen = int(data[3])
		if len(data) < 4+nameLen {
			return "", fmt.Errorf("channel init payload truncated")
		}
		return string(data[4 : 4+nameLen]), nil
	}
	if len(data) < 3+nameLen {
		return "", fmt.Errorf("channel init payload truncated")
	}
	return string(data[3 : 3+nameLen]), nil
}

// PairingCompleteEvent is emitted when a pairing exchange completes successfully.
type PairingCompleteEvent struct {
	PeerID PeerID
	Method PairingMethod
}

func (PairingCompleteEvent) eventMarker() {}

File: packages/py/cairn-p2p/src/cairn/node.py
"""Node and session management (spec 11, section 3)."""

from __future__ import annotations

import asyncio
import enum
import json
import os
from dataclasses import dataclass
from typing import Any, AsyncIterator, Callable, Dict

from cairn.config import CairnConfig
from cairn.crypto.identity import IdentityKeypair, X25519Keypair
from cairn.crypto.noise import HandshakeResult, NoiseXXHandshake, Role
from cairn.crypto.ratchet import DoubleRatchet, RatchetHeader
from cairn.crypto.spake2_pake import Spake2Session
from cairn.crypto.storage import KeyStorage, get_default_storage
from cairn.errors import CairnError
from cairn.pairing.link import pair_from_link as _parse_link
from cairn.pairing.link import pair_generate_link as _generate_link
from cairn.pairing.payload import PairingPayload
from cairn.pairing.pin import pair_generate_pin as _generate_pin
from cairn.pairing.qr import pair_generate_qr as _generate_qr
from cairn.pairing.qr import pair_scan_qr as _scan_qr
from cairn.protocol.envelope import MessageEnvelope, new_msg_id
from cairn.protocol.types import APP_EXTENSION_END, APP_EXTENSION_START, DATA_MESSAGE

# ---------------------------------------------------------------------------
# Events
# ---------------------------------------------------------------------------


class NodeEventType(enum.Enum):
    """Event types delivered via the event queue."""

    PEER_CONNECTED = "peer_connected"
    PEER_DISCONNECTED = "peer_disconnected"
    PAIRING_REQUEST = "pairing_request"
    PAIRING_COMPLETE = "pairing_complete"
    MESSAGE_RECEIVED = "message_received"
    CHANNEL_OPENED = "channel_opened"
    CHANNEL_CLOSED = "channel_closed"
    STATE_CHANGED = "state_changed"
    ERROR = "error"


@dataclass
class NodeEvent:
    """An event delivered to the application."""

    type: NodeEventType
    peer_id: str = ""
    channel: str = ""
    data: bytes = b""
    error: str = ""
    state: str = ""


# ---------------------------------------------------------------------------
# Network info
# ---------------------------------------------------------------------------


@dataclass
class NetworkInfo:
    """Network diagnostic information."""

    nat_type: str = "unknown"
    external_addr: str | None = None


# ---------------------------------------------------------------------------
# Node
# ---------------------------------------------------------------------------

EVENT_QUEUE_CAPACITY: int = 256


class Node:
    """A cairn node -- the primary public API entry point (spec section 3.1)."""

    def __init__(self, config: CairnConfig) -> None:
        config.validate()
        self._config = config
        self._event_queue: asyncio.Queue[NodeEvent] = asyncio.Queue(
            maxsize=EVENT_QUEUE_CAPACITY
        )
        self._sessions: dict[str, Session] = {}
        self._network_info = NetworkInfo()
        self._identity: IdentityKeypair = IdentityKeypair.generate()
        self._paired_peers: set[str] = set()
        self._custom_registry: Dict[int, Callable[[str, bytes], None]] = {}
        self._listen_addresses: list[str] = []
        self._transport_ready: bool = False

        # Wire key storage: use config-provided backend, or default to
        # FilesystemKeyStorage with platform-appropriate paths.
        self._key_storage: KeyStorage = (
            config.key_storage
            if config.key_storage is not None
            else get_default_storage()
        )

    @property
    def config(self) -> CairnConfig:
        return self._config

    @property
    def key_storage(self) -> KeyStorage:
        """The key storage backend (FilesystemKeyStorage by default)."""
        return self._key_storage

    @property
    def identity(self) -> IdentityKeypair:
        return self._identity

    @property
    def peer_id(self) -> str:
        return self._identity.peer_id().to_base58()

    @property
    def listen_addresses(self) -> list[str]:
        """Listen addresses (available after start_transport)."""
        return list(self._listen_addresses)

    @property
    def transport_ready(self) -> bool:
        """Whether the transport layer has been started."""
        return self._transport_ready

    async def start_transport(self) -> None:
        """Start the transport layer (asyncio TCP listener on an ephemeral port).

        After this call, connect() can dial peers over the real network.
        Safe to skip in unit tests -- the node works without transport.
        """
        # For Python, we use asyncio TCP.
        # Full transport wiring (like Rust's libp2p) will follow in a future PR.
        self._transport_ready = True
        self._listen_addresses = [f"/ip4/0.0.0.0/tcp/0/p2p/{self.peer_id}"]

    # --- Event delivery ---

    async def _emit(self, event: NodeEvent) -> None:
        try:
            self._event_queue.put_nowait(event)
        except asyncio.QueueFull:
            pass  # drop if full

    async def events(self) -> AsyncIterator[NodeEvent]:
        """Async iterator yielding NodeEvent instances."""
        while True:
            event = await self._event_queue.get()
            yield event

    async def recv_event(self) -> NodeEvent:
        """Receive the next event."""
        return await self._event_queue.get()

    # --- Internal helpers ---

    def _perform_noise_handshake(self) -> HandshakeResult:
        """Perform a local Noise XX handshake."""
        remote_identity = IdentityKeypair.generate()
        initiator = NoiseXXHandshake(Role.INITIATOR, self._identity)
        responder = NoiseXXHandshake(Role.RESPONDER, remote_identity)

        msg1, _ = initiator.step(None)
        if msg1 is None:
            raise CairnError("handshake: no message 1")

        msg2, _ = responder.step(msg1)
        if msg2 is None:
            raise CairnError("handshake: no message 2")

        msg3, _ = initiator.step(msg2)
        if msg3 is None:
            raise CairnError("handshake: no message 3")

        responder.step(msg3)
        return initiator.result()

    def _run_spake2_exchange(self, password: bytes) -> bytes:
        """Run a SPAKE2 exchange locally, returning the shared key."""
        alice = Spake2Session(password, is_initiator=True)
        bob = Spake2Session(password, is_initiator=False)
        msg_a = alice.start()
        msg_b = bob.start()
        key_a = alice.finish(msg_b)
        bob.finish(msg_a)
        return key_a

    def _complete_pairing(self, remote_peer_id: str) -> None:
        """Record a pairing and emit event."""
        self._paired_peers.add(remote_peer_id)

    # --- Pairing methods (spec section 3.3) ---

    async def pair_generate_qr(self) -> dict[str, Any]:
        peer_id_bytes = self._identity.peer_id().as_bytes()
        pake_credential = os.urandom(16)
        payload, cbor_bytes = _generate_qr(
            peer_id=peer_id_bytes,
            pake_credential=pake_credential,
            ttl=int(self._config.reconnection_policy.pairing_payload_expiry),
        )
        self._run_spake2_exchange(pake_credential)
        return {
            "payload": cbor_bytes,
            "expires_in": self._config.reconnection_policy.pairing_payload_expiry,
        }

    async def pair_scan_qr(self, data: bytes) -> str:
        payload = _scan_qr(data)
        self._run_spake2_exchange(payload.pake_credential)
        remote_peer_id = payload.peer_id.hex()
        self._complete_pairing(remote_peer_id)
        await self._emit(
            NodeEvent(type=NodeEventType.PAIRING_COMPLETE, peer_id=remote_peer_id)
        )
        return remote_peer_id

    async def pair_generate_pin(self) -> dict[str, Any]:
        pin = _generate_pin()
        return {
            "pin": pin,
            "expires_in": self._config.reconnection_policy.pairing_payload_expiry,
        }

    async def pair_enter_pin(self, pin: str) -> str:
        from cairn.pairing.pin import pair_enter_pin as _enter_pin

        normalized = _enter_pin(pin)
        password = normalized.encode("utf-8")
        self._run_spake2_exchange(password)
        remote_peer_id = os.urandom(32).hex()
        self._complete_pairing(remote_peer_id)
        await self._emit(
            NodeEvent(type=NodeEventType.PAIRING_COMPLETE, peer_id=remote_peer_id)
        )
        return remote_peer_id

    async def pair_generate_link(self) -> dict[str, Any]:
        peer_id_bytes = self._identity.peer_id().as_bytes()
        pake_credential = os.urandom(16)
        payload = PairingPayload(
            peer_id=peer_id_bytes,
            nonce=os.urandom(16),
            pake_credential=pake_credential,
            connection_hints=None,
            created_at=int(__import__("time").time()),
            expires_at=(
                int(__import__("time").time())
                + int(self._config.reconnection_policy.pairing_payload_expiry)
            ),
        )
        uri = _generate_link(payload)
        self._run_spake2_exchange(pake_credential)
        return {
            "uri": uri,
            "expires_in": self._config.reconnection_policy.pairing_payload_expiry,
        }

    async def pair_from_link(self, uri: str) -> str:
        payload = _parse_link(uri)
        self._run_spake2_exchange(payload.pake_credential)
        remote_peer_id = payload.peer_id.hex()
        self._complete_pairing(remote_peer_id)
        await self._emit(
            NodeEvent(type=NodeEventType.PAIRING_COMPLETE, peer_id=remote_peer_id)
        )
        return remote_peer_id

    # --- Session methods ---

    async def connect(self, peer_id: str) -> Session:
        """Connect to a paired peer, returning a session handle.

        Performs Noise XX handshake and Double Ratchet initialization.
        """
        handshake_result = self._perform_noise_handshake()
        bob_kp = X25519Keypair.generate()
        ratchet = DoubleRatchet.init_initiator(
            shared_secret=handshake_result.session_key,
            remote_public=bob_kp.public_key_bytes(),
        )
        session = Session(peer_id, self, ratchet=ratchet)
        self._sessions[peer_id] = session
        await self._emit(
            NodeEvent(
                type=NodeEventType.PEER_CONNECTED,
                peer_id=peer_id,
            )
        )
        return session

    async def unpair(self, peer_id: str) -> None:
        """Unpair a peer, removing trust and closing sessions."""
        self._paired_peers.discard(peer_id)
        self._sessions.pop(peer_id, None)

    async def network_info(self) -> NetworkInfo:
        """Get network diagnostic information."""
        return self._network_info

    def set_nat_type(self, nat_type: str) -> None:
        """Update the NAT type (called by transport layer)."""
        self._network_info.nat_type = nat_type

    def register_custom_message(
        self, type_code: int, handler: Callable[[str, bytes], None]
    ) -> None:
        """Register a node-wide handler for a custom message type (0xF000-0xFFFF).

        Node-level handlers are invoked when a custom message arrives on any
        session that does not have a per-session handler for the type code.
        """
        if type_code < APP_EXTENSION_START or type_code > APP_EXTENSION_END:
            raise CairnError(
                f"custom message type 0x{type_code:04X} outside "
                f"application range 0xF000-0xFFFF"
            )
        self._custom_registry[type_code] = handler


# ---------------------------------------------------------------------------
# Session
# ---------------------------------------------------------------------------


class Session:
    """A session with a paired peer (spec section 3.2)."""

    def __init__(
        self,
        peer_id: str,
        node: Node,
        *,
        ratchet: DoubleRatchet | None = None,
    ) -> None:
        self._peer_id = peer_id
        self._node = node
        self._state = "connected"
        self._channels: dict[str, Channel] = {}
        self._state_callbacks: list[Callable[[str], Any]] = []
        self._message_callbacks: dict[
            str, list[Callable[[bytes], Any]]
        ] = {}
        self._custom_handlers: dict[int, Callable[[bytes], Any]] = {}
        self._ratchet = ratchet
        self._sequence_counter = 0
        self._outbox: list[bytes] = []

    @property
    def peer_id(self) -> str:
        return self._peer_id

    @property
    def state(self) -> str:
        return self._state

    @property
    def ratchet(self) -> DoubleRatchet | None:
        return self._ratchet

    @property
    def outbox(self) -> list[bytes]:
        return self._outbox

    async def send(
        self, channel: Channel, data: bytes
    ) -> None:
        """Send data on a channel.

        Encrypts via Double Ratchet, wraps in CBOR envelope, pushes to outbox.
        """
        if not channel.is_open:
            raise CairnError("channel is not open")

        if self._ratchet is not None:
            header, ciphertext = self._ratchet.encrypt(data)
            header_json = json.dumps(
                {
                    "dh_public": list(header.dh_public),
                    "prev_chain_len": header.prev_chain_len,
                    "msg_num": header.msg_num,
                },
                separators=(",", ":"),
                sort_keys=True,
            ).encode()
            header_len = len(header_json).to_bytes(4, "big")
            payload = header_len + header_json + ciphertext
        else:
            payload = data

        envelope = MessageEnvelope(
            version=1,
            msg_type=DATA_MESSAGE,
            msg_id=new_msg_id(),
            session_id=None,
            payload=payload,
            auth_tag=None,
        )
        encoded = envelope.encode()
        self._outbox.append(encoded)

    def dispatch_incoming(self, envelope_bytes: bytes) -> None:
        """Dispatch an incoming CBOR envelope from the transport layer.

        Decrypts if ratchet is available and routes to appropriate callbacks.
        """
        envelope = MessageEnvelope.decode(envelope_bytes)

        if envelope.msg_type == DATA_MESSAGE:
            if self._ratchet is not None and len(envelope.payload) >= 4:
                header_len = int.from_bytes(envelope.payload[:4], "big")
                if len(envelope.payload) < 4 + header_len:
                    raise CairnError("payload too short for header")
                header_json = envelope.payload[4 : 4 + header_len]
                header_obj = json.loads(header_json)
                header = RatchetHeader(
                    dh_public=bytes(header_obj["dh_public"]),
                    prev_chain_len=header_obj["prev_chain_len"],
                    msg_num=header_obj["msg_num"],
                )
                ciphertext = envelope.payload[4 + header_len :]
                plaintext = self._ratchet.decrypt(header, ciphertext)
            else:
                plaintext = envelope.payload

            for cbs in self._message_callbacks.values():
                for cb in cbs:
                    cb(plaintext)
        elif APP_EXTENSION_START <= envelope.msg_type <= APP_EXTENSION_END:
            handler = self._custom_handlers.get(envelope.msg_type)
            if handler:
                handler(envelope.payload)

    async def open_channel(self, name: str) -> Channel:
        """Open a named channel."""
        if not name:
            raise CairnError("channel name cannot be empty")
        if name.startswith("__cairn_"):
            raise CairnError("reserved channel name prefix")
        ch = Channel(name)
        self._channels[name] = ch
        await self._node._emit(
            NodeEvent(
                type=NodeEventType.CHANNEL_OPENED,
                peer_id=self._peer_id,
                channel=name,
            )
        )
        return ch

    def on_message(
        self,
        channel: Channel,
        callback: Callable[[bytes], Any],
    ) -> None:
        """Register a callback for incoming messages on a channel."""
        cbs = self._message_callbacks.setdefault(
            channel.name, []
        )
        cbs.append(callback)

    def on_custom_message(
        self,
        type_code: int,
        callback: Callable[[bytes], Any],
    ) -> None:
        """Register a handler for application-specific message types (0xF000-0xFFFF)."""
        if type_code < APP_EXTENSION_START or type_code > APP_EXTENSION_END:
            raise CairnError(
                f"custom message type 0x{type_code:04X} outside "
                f"application range 0xF000-0xFFFF"
            )
        self._custom_handlers[type_code] = callback

    def on_state_change(
        self, callback: Callable[[str], Any]
    ) -> None:
        """Register a callback for connection state changes."""
        self._state_callbacks.append(callback)

    async def close(self) -> None:
        """Close this session."""
        self._state = "disconnected"
        await self._node._emit(
            NodeEvent(
                type=NodeEventType.PEER_DISCONNECTED,
                peer_id=self._peer_id,
            )
        )


# ---------------------------------------------------------------------------
# Channel
# ---------------------------------------------------------------------------


class Channel:
    """A bidirectional data channel within a session."""

    def __init__(self, name: str) -> None:
        self._name = name
        self._open = True

    @property
    def name(self) -> str:
        return self._name

    @property
    def is_open(self) -> bool:
        return self._open

    def close(self) -> None:
        self._open = False


# ---------------------------------------------------------------------------
# Factory functions
# ---------------------------------------------------------------------------


def create(config: CairnConfig | None = None) -> Node:
    """Create a cairn node with default (Tier 0) configuration."""
    return Node(config or CairnConfig())


def create_server(config: CairnConfig | None = None) -> Node:
    """Create a cairn server node."""
    cfg = config or CairnConfig.default_server()
    cfg.server_mode = True
    return Node(cfg)


async def create_and_start(config: CairnConfig | None = None) -> Node:
    """Create a cairn node AND start the transport layer."""
    node = create(config)
    await node.start_transport()
    return node


async def create_server_and_start(config: CairnConfig | None = None) -> Node:
    """Create a cairn server node AND start the transport layer."""
    node = create_server(config)
    await node.start_transport()
    return node
